src/Kestrel.Transport.Abstractions/Internal/TransportConnection.FeatureCollection.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Net;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Connections.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.Internal
+{
+    public partial class TransportConnection : IFeatureCollection,
+                                               IHttpConnectionFeature,
+                                               IConnectionIdFeature,
+                                               IConnectionTransportFeature,
+                                               IConnectionItemsFeature,
+                                               IMemoryPoolFeature,
+                                               IApplicationTransportFeature,
+                                               ITransportSchedulerFeature
+    {
+        // NOTE: When feature interfaces are added to or removed from this TransportConnection class implementation,
+        // then the list of `features` in the generated code project MUST also be updated.
+        // See also: tools/CodeGenerator/TransportConnectionFeatureCollection.cs
+
+        bool IFeatureCollection.IsReadOnly => false;
+
+        int IFeatureCollection.Revision => _featureRevision;
+
+        object IFeatureCollection.this[Type key]
+        {
+            get
+            {
+                return this[key];
+            }
+            set
+            {
+                this[key] = value;
+            }
+        }
+
+        TFeature IFeatureCollection.Get<TFeature>()
+        {
+            return Get<TFeature>();
```

Comment:
Why introduce a double call?

Just add `IFeatureCollection.` to the private ones in the generated as before? This call can't be inlined as its an interface, and the next can't be inlined as its a its too large (and has too many generic type casts); so its adding a double call cost vs single call cost.

Reply:
The idea is to share the code to generate HttpProtocol which has connection feature fallback and TransportConnection which doesn't have any fallback.